onnxruntime/core/providers/rocm/tunable/gemm_hipblaslt.h

```diff
@@ -0,0 +1,207 @@
+    HIPBLASLT_RETURN_IF_ERROR(hipblasLtMatrixLayoutSetAttribute(
+        mat_c, HIPBLASLT_MATRIX_LAYOUT_BATCH_COUNT, &batch, sizeof(batch)));
+    HIPBLASLT_RETURN_IF_ERROR(hipblasLtMatrixLayoutSetAttribute(
+        mat_c, HIPBLASLT_MATRIX_LAYOUT_STRIDED_BATCH_OFFSET, &stride_c, sizeof(stride_c)));
+  }
+
+  hipblasOperation_t trans_a = (params->opb == BlasOp::N) ? HIPBLAS_OP_N : HIPBLAS_OP_T;
+  hipblasOperation_t trans_b = (params->opa == BlasOp::N) ? HIPBLAS_OP_N : HIPBLAS_OP_T;
+  HIPBLASLT_RETURN_IF_ERROR(hipblasLtMatmulDescSetAttribute(
+      matmul, HIPBLASLT_MATMUL_DESC_TRANSA, &trans_a, sizeof(int32_t)));
+  HIPBLASLT_RETURN_IF_ERROR(hipblasLtMatmulDescSetAttribute(
+      matmul, HIPBLASLT_MATMUL_DESC_TRANSB, &trans_b, sizeof(int32_t)));
+
+  hipblasLtEpilogue_t epilogue;
+  switch (activation_type) {
+    case ActivationType::NONE:
+      epilogue = enable_bias ? HIPBLASLT_EPILOGUE_BIAS : HIPBLASLT_EPILOGUE_DEFAULT;
+      break;
+    case ActivationType::RELU:
+      epilogue = enable_bias ? HIPBLASLT_EPILOGUE_RELU_BIAS : HIPBLASLT_EPILOGUE_RELU;
+      break;
+    case ActivationType::GELU:
+      epilogue = enable_bias ? HIPBLASLT_EPILOGUE_GELU_BIAS : HIPBLASLT_EPILOGUE_GELU;
+      break;
+  }
+  HIPBLASLT_RETURN_IF_ERROR(hipblasLtMatmulDescSetAttribute(
+      matmul, HIPBLASLT_MATMUL_DESC_EPILOGUE, &epilogue, sizeof(epilogue)));
+
+  if (enable_bias) {
+    HIPBLASLT_RETURN_IF_ERROR(hipblasLtMatmulDescSetAttribute(
+        matmul, HIPBLASLT_MATMUL_DESC_BIAS_POINTER, &d_bias, sizeof(void*)));
+  }
+  if (enable_scaleD) {
+    HIPBLASLT_RETURN_IF_ERROR(hipblasLtMatmulDescSetAttribute(
+        matmul, HIPBLASLT_MATMUL_DESC_D_SCALE_POINTER, &d_scaleD, sizeof(void*)));
+  }
+
+  hipblasLtMatmulPreference_t pref;
+  void* workspace;
+  size_t max_workspace_size = 32 * 1024 * 1024;
+  HIPBLASLT_RETURN_IF_ERROR(hipblasLtMatmulPreferenceCreate(&pref));
+  HIPBLASLT_RETURN_IF_ERROR(hipblasLtMatmulPreferenceSetAttribute(
+      pref, HIPBLASLT_MATMUL_PREF_MAX_WORKSPACE_BYTES, &max_workspace_size, sizeof(max_workspace_size)));
+
+  const int heuristic_result_count = 3;
+  hipblasLtMatmulHeuristicResult_t heuristic_result[heuristic_result_count] = {0};
+  int ret_algo_count = 0;
+  HIPBLASLT_RETURN_IF_ERROR(hipblasLtMatmulAlgoGetHeuristic(handle,
+                                                            matmul,
+                                                            mat_a,
+                                                            mat_b,
+                                                            mat_c,
+                                                            mat_c,
+                                                            pref,
+                                                            heuristic_result_count,
+                                                            heuristic_result,
+                                                            &ret_algo_count));
+
```

Comment:
> &ret_algo_count

What if `hipblasLtMatmulAlgoGetHeuristic` sets `ret_algo_count` to 0? This isn't being checked anywhere. Also, we're only using heuristic_result[0], why bother passing in space for three?

Reply:
> What if hipblasLtMatmulAlgoGetHeuristic sets ret_algo_count to 0? This isn't being checked anywhere.

hipBLASLt guarantees that the returned algo_count is at least one. Added an assert here.

> we're only using heuristic_result[0], why bother passing in space for three?

You are right, since we are only taking the first heuristic result, the remaining two will be unnecessary. I changed it to 1 but kept the logic here in case we need more to select from in the future.